flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala

```diff
@@ -1248,72 +1241,125 @@ object ScalarOperatorGens {
-          s"""
-             |$DECIMAL_UTIL.castFrom(
-             |  ((double) ($operandTerm.getMillisecond() / 1000.0)),
-             |  ${dt.getPrecision}, ${dt.getScale})
-           """.stripMargin
-      }
-
-    // Tinyint -> Timestamp
-    // Smallint -> Timestamp
-    // Int -> Timestamp
-    // Bigint -> Timestamp
+    // Disable cast conversion between Numeric type and  Timestamp type
     case (TINYINT, TIMESTAMP_WITHOUT_TIME_ZONE) |
          (SMALLINT, TIMESTAMP_WITHOUT_TIME_ZONE) |
          (INTEGER, TIMESTAMP_WITHOUT_TIME_ZONE) |
-         (BIGINT, TIMESTAMP_WITHOUT_TIME_ZONE) =>
+         (BIGINT, TIMESTAMP_WITHOUT_TIME_ZONE) |
+         (FLOAT, TIMESTAMP_WITHOUT_TIME_ZONE) |
+         (DOUBLE, TIMESTAMP_WITHOUT_TIME_ZONE) |
+         (DECIMAL, TIMESTAMP_WITHOUT_TIME_ZONE) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, TINYINT) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, SMALLINT) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, INTEGER) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, BIGINT) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, FLOAT) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, DOUBLE) |
+         (TIMESTAMP_WITHOUT_TIME_ZONE, DECIMAL) => {
+      if (TIMESTAMP_WITHOUT_TIME_ZONE.equals(targetType.getTypeRoot)) {
+        throw new ValidationException("The cast conversion from NUMERIC type to TIMESTAMP type" +
+          " is not allowed, it's recommended to use TO_TIMESTAMP(FROM_UNIXTIME(numeric_col))" +
+          " instead, note the numeric is in seconds.")
+      } else {
+        throw new ValidationException("The cast conversion from TIMESTAMP type to NUMERIC type" +
+          " is not allowed,it's recommended to use" +
+          " UNIX_TIMESTAMP(CAST(timestamp_col AS STRING)) instead.")
+      }
+    }
+
+    // Tinyint -> TimestampLtz
+    // Smallint -> TimestampLtz
+    // Int -> TimestampLtz
+    // Bigint -> TimestampLtz
+    case (TINYINT, TIMESTAMP_WITH_LOCAL_TIME_ZONE) |
+         (SMALLINT, TIMESTAMP_WITH_LOCAL_TIME_ZONE) |
+         (INTEGER, TIMESTAMP_WITH_LOCAL_TIME_ZONE) |
+         (BIGINT, TIMESTAMP_WITH_LOCAL_TIME_ZONE) =>
       generateUnaryOperatorIfNotNull(ctx, targetType, operand) {
         operandTerm => s"$TIMESTAMP_DATA.fromEpochMillis(((long) $operandTerm) * 1000)"
       }
 
-    // Float -> Timestamp
-    // Double -> Timestamp
-    case (FLOAT, TIMESTAMP_WITHOUT_TIME_ZONE) |
-         (DOUBLE, TIMESTAMP_WITHOUT_TIME_ZONE) =>
+    // Float -> TimestampLtz
+    // Double -> TimestampLtz
+    case (FLOAT, TIMESTAMP_WITH_LOCAL_TIME_ZONE) |
+         (DOUBLE, TIMESTAMP_WITH_LOCAL_TIME_ZONE) =>
+      generateUnaryOperatorIfNotNull(ctx, targetType, operand) {
+        operandTerm =>
+          s"""
+             |$DECIMAL_UTIL.castToTimestamp(
+             |  $DECIMAL_UTIL.castFrom(
+             |    (double) $operandTerm,
+             |    $DECIMAL_UTIL.DECIMAL_SYSTEM_DEFAULT.getPrecision(),
+             |    $DECIMAL_UTIL.DECIMAL_SYSTEM_DEFAULT.getScale()))
+             """.stripMargin
```

Comment:
Because the operand is a floating point number, I think it's fine to just calculate on it. Using decimal is a little heavy. I was just concerned about the overflowing. We can avoid overflowing in the following way:

```
TimestampData.fromEpochMillis(
    operandTerm * 1000, 
    (operandTerm - (int) operandTerm) * 1000_000_000 % 1000_000);
```